File: python/Baekjoon/1654.py
import sys

k, n = map(int, sys.stdin.readline().rstrip().split())
lans = []
for i in range(k):
    lans.append(int(sys.stdin.readline().rstrip()))
result = 0

start = 1
end = sum(lans) // n

while start <= end:
    mid = (start + end) // 2
    count = sum(i // mid for i in lans)

    if count >= n:
        result = mid
        start = mid + 1
    else: end = mid - 1

print(result)

File: Baekjoon/1654.js
const readline = require("readline");

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let [k, n] = [0, 0];
const lans = [];
let lineCount = 0;

rl.on("line", (input) => {
  if (lineCount == 0) [k, n] = input.split(" ").map(Number);
  else lans.push(+input);
  lineCount++;
});

rl.on("close", () => {
  let start = 1;
  let end = Math.max(...lans);

  while (start <= end) {
    let mid = parseInt((start + end) / 2);
    let count = 0;

    lans.forEach((lan) => {
      count += parseInt(lan / mid);
    });

    if (count < n) end = mid - 1;
    else start = mid + 1;
  }

  console.log(end);

  process.exit();
});
